include/envoy/http/conn_pool.h

```diff
@@ -48,7 +48,7 @@ class Callbacks {
 /**
  * An instance of a generic connection pool.
  */
-class Instance : public Envoy::ConnectionPool::Instance {
+class Instance : virtual public Envoy::ConnectionPool::Instance, public Event::DeferredDeletable {
```

Comment:
using new names

So the shared connection pool interface has the drain functions.  this inherits (so we know that the http interface can drain), and the HttpConnectionPoolBase inherits (because it implements drain), but then HttpConnectionPoolBase inherits from the http interface, and creates the diamond inheritance pattern which AFIK can only be solved with virtual inheritance (though correct me if I'm wrong, or there's a more elegant way to  handle it)

Reply:
I'm not sure if this will always be the case, but `Envoy::ConnectionPool::ConnPoolImplBase` doesn't implement or use any of the methods defined in `Envoy::ConnectionPool::Instance`, so it doesn't need to inherit from it at all (I don't think). Then there's no diamond.

Reply:
Yeah, I had thought moving the functions to the http specific subclass would be sufficient but the compiler disagreed with me (envoy-presubmit (Linux-x64 gcc failure due to disambuation issues).  The second upload (adding virtual here) fixed.

https://github.com/envoyproxy/envoy/pull/11689/commits/a0b14b77342ac5fef71247ddb9fd72a17f25f65d

Reply:
I did a test and made `ConnPoolImplBase` not inheirt from `ConnectionPool::Instance` at all, and made this line not virtual inheritance, and it compiled (linux, clang).

Reply:
right, it passed all the builds other than the one mentioned.  I can revert it and we can look at CI again...